compiler/optimizer/AsyncCheckInsertion.cpp

```diff
@@ -102,7 +119,18 @@ bool TR_AsyncCheckInsertion::shouldPerform()
    if (comp()->isProfilingCompilation() || comp()->generateArraylets())
       return false;
 
-   if (comp()->getOption(TR_EnableOSR))  // cannot move async checks around arbitrarily if OSR is possible
+   // It is not safe to add an asynccheck under involuntary OSR
+   // as a transition may have to occur at the added point and the
+   // required infrastructure may not exist
+   //
+   if (comp()->getOption(TR_EnableOSR) && comp()->getOSRMode() == TR::involuntaryOSR)
+      return false;
+
+   // Should not be performed at very hot or scorching. Sampling does
+   // not matter at these points, as recompilation will not occur based
+   // on it.
+   //
+   if (comp()->getMethodHotness() >= veryHot)
```

Comment:
Probably not - I wasn't aware of non-profiling very hot.

This test isn't really necessary anyway because higher opt levels use `redundantAsyncCheckRemoval` instead. If you like I can delete it to let the strategy decide, or change it to be more precise, [perhaps something like this](https://github.com/eclipse/omr/blob/3a5b9f9b193b68cad98a07421aa34f6df6ae1614/compiler/optimizer/RedundantAsyncCheckRemoval.cpp#L157-L164).